api/wasm/cpp/proxy_wasm_common.h

```diff
@@ -73,7 +73,8 @@ enum class BufferType : int32_t {
   NetworkUpstreamData = 3,   // During the onLog callback these are immutable
   HttpCallResponseBody = 4,  // Immutable
   GrpcReceiveBuffer = 5,     // Immutable
-  MAX = 5,
+  BufferedBody = 6,          // After returning StopIterationAndBuffer
```

Comment:
I'm strongly against introducing another enum for this purpose.

Why this buffer cannot be returned using `getBuffer(HttpRequestBody)` and `getBuffer(HttpResponseBody)`?

Reply:
If a filter returns StopIterationAndBuffer from encode/decodeData, then Envoy gathers all the data in each chunk into a single buffer. It also enforces a maximum buffer length and rejects the request or fails the response if it's too big. That buffer is available by calling "decodingBuffer" or "encodingBuffer" on the filter callbacks, and this PR makes it possible to call those functions. (There's another function used to modify it, and we'll get to that later.) From what I can tell, the Lua and DynamoDb filters seem to use this capability.

Without a way to access this buffer, then a filter that wants to look at all the body data at once has to buffer the data itself by aggregating all buffers passed to encode/decodeData, and handle its own logic to prevent running out of memory. Or, the filter writer has to tell users to always include the special "buffer" filter before their filter.

Reply:
Also -- if you want to have a longer conversation...

If one is trying to write a filter that validates the message body and either rejects or accepts it (for instance, it wants to parse some JSON) or if it wants to read a message body and then write out a new one (for instance, it wants to then transform that JSON) then we need some way for the filter to safely and efficiently buffer the body and change it later.

However, unless people are going to use the WASM support to build lower-level filters to do things like compression, then we don't need all the complexity of the full Envoy API -- we could instead design something with a  simpler interface. I have some ideas on that and I'd be happy to propose one if you'd like, but I assumed that the goal if this project was to offer the same flexibility to WASM authors that Envoy offers C++ programmers.

Reply:
@gbrail sorry if my response wasn't clear, I definitely want to have access to the buffered data for obvious reasons. What I'm suggesting is that this data should be available via `getBuffer(HttpRequestBody)` call (i.e. consumer shouldn't care whether this is new or buffered data) and not via a separate call. Does it make sense?